Create a page for policies and guidelines for maintainers (#20)
neoforged/governance: maintainer-policies-and-guidelines.md

# Maintainer Policies and Guidelines

This document aims to clarify hard policies that Maintainers have to follow as well as less-enforced guidelines that are recommended to be followed. The goal is to help Maintainers understand what is expected of them and what actions should be done to reduce the chances of conflict within the Maintainer team.

## Policies

> - `c` tags in 1.21+ NeoForge should be kept in sync with Fabric's `c` tags as close as possible. New `c` tags or changes to existing `c` tag should by ran by Fabric to ensure they will agree to make the same changes. `c` tags are to be a standard between loaders with Fabric being our main partner in this effort.

Reviewer: This is not something important enough to have in this document IMO.

Author: If we don't write it down here, where do we document that we have to keep c tags in sync with fabric? It's currently an unspoken rule and if you and I were to get hit by a bus, how do we know the next maintainer wont start changing c tags without checking with fabric?

>   - Failure to meet this expectation will result in the change being undone so NeoForge matches Fabric again and the incident reviewed to understand how we failed to communicate with Fabric.

Reviewer: This reads very punitive, and adds little value. Yes mistakes do happen and we will try to correct them. An immediate revert may or may not be the solution, especially if the mistake is only noticed later. To be determined on a case-by-case basis, IMO.

## Guidelines

- When making a "big-deal" PR, please post a link to the PR in the NeoForge Discord's maintenance-talk channel, ping the maintainer role, and leave a short sentence on what the PR is about.

  - A "big-deal" PR is one that has a significant impact on many systems, changes the project ecosystem, or requires significant coordination with external projects.
  

>     - An example of a significant impact on many systems, Transfer PR which involved creating a transaction system and refactoring the item/fluid/energy handlers. Many systems affected and massive amount of changes. This is a big deal.

Reviewer: We had many discussions on this, both public and internal, and there is a dedicated brainstorming topic to which all maintainers got added. If anyone is not aware by now I don't know what to say.

Author: Transfer PR is just an example of a big deal PR Not an example of a miscommunication. Well, I guess the second Transfer PR being made could be seen as a miscommunication but I'm using Adrian's/Soaryn's PR here as one that is a Big Deal and other PRs of similar magnitude should have a Maintainer ping. 

Sorry if that wasn't clear. These are not examples of miscommunications. They are examples of big deal stuff

>     - An example of a project ecosystem change is the PR to merge ModLauncher code into FancyModLoader to remove the dependency to ModLauncher. We folded one deliberately-external project into our own project. This is a big deal as it changes what is in scope of the NeoForged organization and changed the relation between projects. If two of our own NeoForged projects were to merge or split up into more, this too would be a big deal as maintainers would definitely like to know as keeping track of what project relates to what is important. *(In addition, when cutting out reliance on an external project that we had close relations with, we should notify the owner of the external project before we cut the ties)*

Reviewer: This feels like trying to carve out a rule to justify why the FML merge should have been more visible. Fair enough, we could have done a better job. But keep in mind that this change has little impact on users of NeoForge, was already discussed many times over the past months, and even had alpha releases on our maven. The only reason why the change had to happen again slightly differently was due to a technical limitation in our current loader code.

Author: See my comment above. Basically trying to say what is a big deal thing and not an example of miscommunications. This also applies to how we switched off of Mixins to Fabric Mixins. I can add that here to make this be 2 examples of cutting out dependencies to 'external' projects we had close ties to

    - Lastly, a solid example of a coordination work is the Unify Tag PR which created a new policy that new `c` tags have to be ran by Fabric first to see if they agree with adding them and requiring NeoForge to keep their `c` tags to be as similar to Fabric's as possible. The Unify Tag PR is also a significant many-system impact example too due to the wide-reaching nature of tags.

>   - There is no hard enforcement of this guideline. Rather it aims to show that communication of a "big-deal" PR is important to helping to keep the team in the loop of what is going on and prevent conflicts between teammates. The examples given helps to give a sense to what kind of PR is considered a "big-deal" but it is left a bit open-ended as PRs can take many forms. We will defer to the PR creator's to use their best judgement on whether a PR is something to keep the team informed about or not. The rate of "big-deal" PRs should be uncommon so ideally, this should not lead to an excess of maintainer pings.

Reviewer: So what is the point of this guideline if it's not enforced?

Author: Not everything has to be enforced for people to understand what is expected of them. 

Case in point, that argument that happened in Discord about the breaking change window. We never wrote down how long we want to do breaking change window, and that it is not a rule that is enforced. We play BCs by ear. But if we wrote down that it is 1 month long and we will ignore hotfix versions, that whole argument would never happened as we can point to the guidelines here and show what we are generally following for breaking changes.

Same thing here. This help new maintainers to come in knowing that they should keep the rest of the team in the loop with big deal PRs. So that they won't repeat the issues we had in past. If they do fail to communicate, that incident will be handled as its own thing based on what happened. Maintainers then will figure out best course of action.

Really, it's not having these actually written out somewhere that I find is generally the problem. It makes it feel like there's no best practice or recommended procedure for a maintainer to keep in mind.

Now, if the rest of the maintainer team really wants this to be enforced, we could make it a rule. But given how big-deal PRs could be a bit subjective, that's where it can be tricky and why I think it is best as a guideline to make sure maintainers are aware other maintainers want to be roped in on big stuff happening.

# Maintainer Policies and Guidelines

> This document aims to clarify hard policies that Maintainers have to follow as well as less-enforced guidelines that are recommended to be followed. The goal is to help Maintainers understand what is expected of them and what actions should be done to reduce the chances of conflict within the Maintainer team.

Reviewer: ```suggestion
This document aims to clarify policies by which the Maintainer team operates. The goal is to help Maintainers understand what is expected of them and what actions should be done to reduce the chances of conflict within the Maintainer team.
```

## Policies

> 

Reviewer: You can add that PRs must not be merged without checking with a maintainer that requested changes and/or got assigned the PR (possibly self-assigned). The review should not be dismissed without first checking with the maintainer. This also serves as a way for maintainers to request others to wait for their review.

Author: Good idea. i recall this coming up but I forgot about that

- `c` tags in 1.21+ NeoForge should be kept in sync with Fabric's `c` tags as close as possible. New `c` tags or changes to existing `c` tag should by ran by Fabric to ensure they will agree to make the same changes. `c` tags are to be a standard between loaders with Fabric being our main partner in this effort.

>   - Failure to meet this expectation will result in incident being reviewed to understand how we failed to communicate with Fabric. The PR in question could be reverted out of NeoForge if Fabric rejects the tag changes. Will be handled on a case-by-case basis.

Reviewer: Remove this paragraph.

## Guidelines

> - Breaking change window for the NeoForge project will generally aim to last only 1 month after a significant Minecraft version.

Reviewer: remove "only", replace by "about".

  - Hotfix Minecraft versions or very small Minecraft versions will not reset the breaking change window.

  - This window is flexible and may be longer if there is significant breaking change PRs that need to be released for that Minecraft version but not yet ready. The 1 month timeframe is just a goal we would like to achieve but we understand it is not always possible or optimal.

> - When making a "big-deal" PR, please post a link to the PR in the NeoForge Discord's maintenance-talk channel, ping the maintainer role, and leave a short sentence on what the PR is about.

Reviewer: This entire paragraph and the sublist can be replaced by something much shorter. For example `Maintainers should ping @Maintainers on discord before merging important PRs. What is considered important enough to warrant a ping is left to maintainer discretion, but it is better to ping too often rather than not enough.`
